Generate pipeline (#334)
openvinotoolkit/openvino.genai: src/README.md

# OpenVINO Generate API

## Usage 

First of all you need to convert your model with optimum-cli
``` sh
optimum-cli export openvino --model "TinyLlama/TinyLlama-1.1B-Chat-v1.0" --weight-format fp16 --trust-remote-code "TinyLlama-1.1B-Chat-v1.0"
pip install openvino-genai
```

`LLMPipeline` is the main object used for decoding. You can construct it straight away from the folder with the converted model. It will automatically load the main model, tokenizer, detokenizer and default generation configuration.

### Python

A minimalist example:
```python
import openvino_genai as ov_genai
pipe = ov_genai.LLMPipeline(model_path, "CPU")

> print(pipe.generate("The Sun is yellow bacause"))

Reviewer: is this misspelling intentional? - bacause"

Author: yes, there are not comma and periods so that network will continue this sentence

Reviewer: yeah it's just the word "because" is spelled differently )) was wondering if this is also a thing for the network to fix)

```

Calling generate with custom generation config parameters, e.g. config for grouped beam search
```python
import openvino_genai as ov_genai
pipe = ov_genai.LLMPipeline(model_path, "CPU")

> result = pipe.generate("The Sun is yellow bacause", max_new_tokens=30, num_groups=3, group_size=5, diversity_penalty=1.5)

Author: Added validation for `GenerationConfig` at the beginning of generation. 

In your example, if default `generatin_config.json` had `do_sample=True`, then user will be able to load pipeline, but it will fail at the beginning of generation with a message to modify conflicting values, e.g. `do_sample=False` or disable beam search.

print(result)
```

output:
```
'it is made up of carbon atoms. The carbon atoms are arranged in a linear pattern, which gives the yellow color. The arrangement of carbon atoms in'
```

A simples chat in python:
```python
import openvino_genai as ov_genai

> pipe = ov_ov_genai.LLMPipeline(model_path)

Reviewer 2: ov_ov_genai -> ov_genai

Author: fixed

> config = {'num_groups': 3, 'group_size': 5, 'diversity_penalty': 1.5}

Author: If `eos_token_id` is missing in `generation_config.json` will take it from tokenizer (which loaded it from wherever possible). If eos_token_id is specified in config then values from tokenizer will not be used, so theoretically can differ as it happens for some model in HF.

> config = {'num_groups': 3, 'group_size': 5, 'diversity_penalty': 1.5}

Reviewer 3: we can take it from default generation config (I mean user config will inherit this value from generation_config.json).

> pipe.set_generation_cofnig(config)

Reviewer 2: I tried and it does not work like this for me, should it work ? It woks for me if config is GenerationConfig object and I got with get_generation_config before

Reviewer 2: Is it possible to add description about options , which we can configure in config ?

    std::string prompt;

    std::string model_path = argv[1];
    ov::genai::LLMPipeline pipe(model_path, "CPU");
    
    ov::genai::GenerationConfig config = pipe.get_generation_config();
    config.max_new_tokens = 256;
    config.num_groups = 3;
    config.group_size = 5;
    config.diversity_penalty = 1.0f;
    
    pipe.start_chat();
    for (;;;) {
        std::cout << "question:\n";
        std::getline(std::cin, prompt);
        if (prompt == "Stop!")
            break;

        std::cout << "answer:\n";
        auto answer = pipe(prompt, config);
        std::cout << answer << std::endl;
    }
    pipe.finish_chat();
}
```

Streaming example with lambda function
``` cpp
#include "openvino/genai/llm_pipeline.hpp"
#include <iostream>

int main(int argc, char* argv[]) {
    std::string model_path = argv[1];
    ov::genai::LLMPipeline pipe(model_path, "CPU");
        
    auto streamer = [](std::string word) { std::cout << word << std::flush; };
    std::cout << pipe.generate("The Sun is yellow bacause", streamer);
}
```

Streaming with a custom class
``` cpp
#include "openvino/genai/streamer_base.hpp"
#include "openvino/genai/llm_pipeline.hpp"
#include <iostream>

> class CustomStreamer: public ov::genai::StreamerBase {

Reviewer 3: what do you think?
@Wovchena @pavel-esir

# OpenVINO Generate API

## Usage 

First of all you need to convert your model with optimum-cli
``` sh
optimum-cli export openvino --model "TinyLlama/TinyLlama-1.1B-Chat-v1.0" --weight-format fp16 --trust-remote-code "TinyLlama-1.1B-Chat-v1.0"
pip install openvino-genai
```

> LLMPipeline is the main object used for decoding. You can initiliza it straigh away from the folder with the converted model. It will automanically load the main model, tokenizer, detokenizer and default generation configuration.

Reviewer 4: ```suggestion
`LLMPipeline` is the main object used for decoding. You can initialize it straight away from the folder with the converted model. It will automatically load the main model, tokenizer, detokenizer and default generation configuration.
```

### Python

A minimalist example:
```python
import openvino_genai as ov_genai
pipe = ov_genai.LLMPipeline(model_path, "CPU")
print(pipe.generate("The Sun is yellow bacause"))
```

> Calling generate with custom generation config parameters, e.g. config for grouped beam search

Reviewer 4: ```suggestion
Calling generate with custom generation config parameters, e.g. config for grouped beam search:
```

```python
import openvino_genai as ov_genai
pipe = ov_genai.LLMPipeline(model_path, "CPU")

> res = pipe.generate("The Sun is yellow bacause", max_new_tokens=30, num_groups=3, group_size=5)

Author: added diversity penalty as well

print(res)
```

output:
```
'it is made up of carbon atoms. The carbon atoms are arranged in a linear pattern, which gives the yellow color. The arrangement of carbon atoms in'
```

> A simples chat in python:

Reviewer 4: ```suggestion
A simple chat in python:
```

    prompt = input()
    if prompt == 'Stop!':
        break
    print(pipe(prompt))
pipe.finish_chat()

```

Test to compare with Huggingface outputs

### C++

Minimalistc example
```cpp
#include "openvino/genai/llm_pipeline.hpp"
#include <iostream>

int main(int argc, char* argv[]) {
    std::string model_path = argv[1];
    ov::LLMPipeline pipe(model_path, "CPU");
    std::cout << pipe.generate("The Sun is yellow bacause");
}
```

Using Group Beam Search Decoding
```cpp
#include "openvino/genai/llm_pipeline.hpp"
#include <iostream>

int main(int argc, char* argv[]) {
    std::string model_path = argv[1];
    ov::LLMPipeline pipe(model_path, "CPU");

    ov::GenerationConfig config = pipe.get_generation_config();
    config.max_new_tokens = 256;
    config.num_groups = 3;
    config.group_size = 5;
    config.diversity_penalty = 1.0f;

    std::cout << pipe.generate("The Sun is yellow bacause", config);
}
```

A simplest chat in C++
``` cpp
#include "openvino/genai/llm_pipeline.hpp"
#include <iostream>

int main(int argc, char* argv[]) {
    std::string prompt;

    std::string model_path = argv[1];
    ov::LLMPipeline pipe(model_path, "CPU");

    pipe.start_chat();

>     for (size_t i = 0; i < questions.size(); i++) {

Author: replaced with reading from the `std::cin`

        std::cout << "question:\n";
        std::getline(std::cin, prompt);

        std::cout << pipe(prompt) << std::endl>>;
    }
    pipe.finish_chat();
}
```

Specifying generation_config to use grouped beam search
``` cpp
int main(int argc, char* argv[]) {
    std::string prompt;

    std::string model_path = argv[1];
    ov::LLMPipeline pipe(model_path, "CPU");
    
    ov::GenerationConfig config = pipe.get_generation_config();
    config.max_new_tokens = 256;
    config.num_groups = 3;
    config.group_size = 5;
    config.diversity_penalty = 1.0f;
    

>     auto streamer = [](std::string word) { std::cout << word << std::flush; };

Author: removed

}
```

Streaming example with lambda function
``` cpp
#include "openvino/genai/llm_pipeline.hpp"
#include <iostream>

int main(int argc, char* argv[]) {
    std::string model_path = argv[1];
    ov::genai::LLMPipeline pipe(model_path, "CPU");
        
    auto streamer = [](std::string word) { std::cout << word << std::flush; };
    std::cout << pipe.generate("The Sun is yellow bacause", streamer);
}
```

Streaming with a custom class
``` cpp
#include "openvino/genai/streamer_base.hpp"
#include "openvino/genai/llm_pipeline.hpp"
#include <iostream>

class CustomStreamer: public ov::genai::StreamerBase {
public:
    bool put(int64_t token) {
        bool stop_flag = false;
        /* custom decoding/tokens processing code
        tokens_cache.push_back(token);
        std::string text = m_tokenizer.decode(tokens_cache);
        ...
        */
        return stop_flag;
    };

    void end() {
        /* custom finalization */
    };
};

int main(int argc, char* argv[]) {
    CustomStreamer custom_streamer;

    std::string model_path = argv[1];
    ov::genai::LLMPipeline pipe(model_path, "CPU");
    std::cout << pipe.generate("The Sun is yellow bacause", custom_streamer);
}
```

> ## How it works

Reviewer 3: @Wovchena @yatarkan  
let's move it to dev docs / how it works.

class CustomStreamer: public ov::genai::StreamerBase {
public:
    bool put(int64_t token) {
        /* custom decoding/tokens processing code
        tokens_cache.push_back(token);
        std::string text = m_tokenizer.decode(tokens_cache);
        ...
        */
    };

>     bool end() {

Author: typo

Author: changed to void
